src/Stack/Init.hs

```diff
@@ -176,65 +177,124 @@ renderStackYaml p ignoredPackages dupPackages =
-                B.byteString "\n"
-
     goOthers o
         | HM.null o = mempty
         | otherwise = assert False $ B.byteString $ Yaml.encode o
 
+    -- Per Section Help
+    comments =
+        [ ("user-message"     , userMsgHelp)
+        , ("resolver"         , resolverHelp)
+        , ("packages"         , packageHelp)
+        , ("extra-deps"       , "# Dependency packages to be pulled from upstream that are not in the resolver\n# (e.g., acme-missiles-0.3)")
+        , ("flags"            , "# Override default flag values for local packages and extra-deps")
+        , ("extra-package-dbs", "# Extra package databases containing global packages")
+        ]
+
+    -- Help strings
+    headerHelp = commentHelp
+        [ "This file was automatically generated by 'stack init'"
+        , ""
+        , "Some commonly used options have been documented as comments in this file."
+        , "For advanced use and comprehensive documentation of the format, please see:"
+        , "http://docs.haskellstack.org/en/stable/yaml_configuration/"
+        ]
+
+    resolverHelp = commentHelp
+        [ "Resolver to choose a 'specific' stackage snapshot or a compiler version."
+        , "A snapshot resolver dictates the compiler version and the set of packages"
+        , "to be used for project dependencies. For example:"
+        , ""
+        , "resolver: lts-3.5"
+        , "resolver: nightly-2015-09-21"
+        , "resolver: ghc-7.10.2"
+        , "resolver: ghcjs-0.1.0_ghc-7.10.2"
+        , "resolver:"
+        , " name: custom-snapshot"
+        , " location: \"./custom-snapshot.yaml\""
+        ]
+
+    userMsgHelp = commentHelp
+        [ "A warning or info to be displayed to the user on config load." ]
+
+    packageHelp = commentHelp
+        [ "User packages to be built."
+        , "Various formats can be used as shown in the example below."
+        , ""
+        , "packages:"
+        , "- some-directory"
+        , "- https://example.com/foo/bar/baz-0.0.2.tar.gz"
+        , "- location:"
+        , "   git: https://github.com/commercialhaskell/stack.git"
+        , "   commit: e7b331f14bcffb8367cd58fbfc8b40ec7642100a"
+        , "- location: https://github.com/commercialhaskell/stack/commit/e7b331f14bcffb8367cd58fbfc8b40ec7642100a"
+        , " subdirs:"
+        , " - auto-update"
+        , " - wai"
+        , ""
+        , "A package marked 'extra-dep: true' will only be built if demanded by a"
+        , "non-dependency (i.e. a user package), and its test suites and benchmarks"
+        , "will not be run. This is useful for tweaking upstream packages."
+        ]
+
+    footerHelp = commentHelp
+        [ "Control whether we use the GHC we find on the path"
+        , "system-ghc: true"
+        , ""
+        , "Require a specific version of stack, using version ranges"
+        , "require-stack-version: -any # Default"
+        , "require-stack-version: >= 1.0.0"
```

Comment:
this one has always bothered me since it is not valid yaml the version range should be quoted :)
this seems to be updated manually with version bumps, but why not just grab it from Paths_stack dropping the minor version?